### PR TITLE
Bug 1116780 - [Messages] We should not focus the composer after clicking a notification. r=julien

### PR DESCRIPTION
**Pull request by azasypkin**

### CODE REVIEW

**Comment by try-server-hook** [comment]
[azasypkin Aleh Zasypkin (azasypkin)](https://github.com/azasypkin) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=dc12171615e5)

File: apps/sms/js/activity_handler.js [hunk 1]
```diff
@@ -125,21 +125,26 @@ var ActivityHandler = {
...
+          return ActivityHandler._findContactByNumber(viewInfo.number)
+            .then((contact) => viewInfo.contact = contact);
+        }
+      )
+      // case no contact and no thread id: gobble the error
+      .catch(() => {});
+    }
+
+    return (threadPromise || Promise.resolve()).then(
+      () => this.toView(viewInfo, !!viewInfo.threadId /* focusComposer */)
```
Comment (julienw) [inline apps/sms/js/activity_handler.js]: for the "focusComposer" boolean, it's quite misleading, because we'll focus also in the case of a "new" that enters the composer (so no threadId); but that's handled by ThreadUI.

I'd say we want to focus if we have a number. And we could even reuse the information in ThreadUI.handleActivity instead of having the same "if".

So I'd do a "var focusComposer = false" before the if, and set it to true inside the if. (or something similar, up to you). And then use it here. And clean up ThreadUI.handleActivity :)

File: apps/sms/js/activity_handler.js [hunk 2]
```diff
@@ -300,62 +305,51 @@ var ActivityHandler = {
...
+   * to preset the compose input with, "contact" is an optional MozContact
+   * instance, "threadId" is an optional threadId corresponding to a new or
+   * existing thread.
+   * @param {Boolean} focusComposer Indicates whether we need to focus composer
+   * when we navigate to Thread panel.
+   */
+  toView: function ah_toView(message, focusComposer) {
     if (!message) {
       return;
     }
```
Comment (julienw) [inline apps/sms/js/activity_handler.js]: so, should we remove this if?

File: apps/sms/js/activity_handler.js [hunk 3]
```diff
@@ -300,62 +305,51 @@ var ActivityHandler = {
...
         return;
       }
 
-      document.addEventListener('visibilitychange',
-        function waitVisibility() {
-          document.removeEventListener('visibilitychange', waitVisibility);
-          showAction();
+      document.addEventListener('visibilitychange', function waitVisibility() {
+        document.removeEventListener('visibilitychange', waitVisibility);
+        navigateToView();
```
Comment (julienw) [inline apps/sms/js/activity_handler.js]: I don't really remember why we wait for a visibility change... but let's not change this in a blocker.

File: apps/sms/test/unit/activity_handler_test.js [hunk 4]
```diff
@@ -734,6 +734,11 @@ suite('ActivityHandler', function() {
   });
 
   suite('"new" activity', function() {
+    function onceNewActivityCompleted() {
+      sinon.assert.called(ActivityHandler._onNewActivity);
+      return ActivityHandler._onNewActivity.firstCall.returnValue;
```
Comment (julienw) [inline apps/sms/test/unit/activity_handler_test.js]: maybe using lastCall is better, in case a test starts 2 activities?

File: apps/sms/js/activity_handler.js (hunk 2 above)
Comment (azasypkin) [inline apps/sms/js/activity_handler.js]: Sure! Just haven't removed it in the second patch yet :)

File: apps/sms/test/unit/activity_handler_test.js (hunk 4 above)
Comment (azasypkin) [inline apps/sms/test/unit/activity_handler_test.js]: Yeah, it'd be safer to use ```lastCall```, thanks

File: apps/sms/js/activity_handler.js (hunk 1 above)
Comment (azasypkin) [inline apps/sms/js/activity_handler.js]: >for the "focusComposer" boolean, it's quite misleading, because we'll focus also in the case of a "new" that enters the composer (so no threadId); but that's handled by ThreadUI.

Yeah, I see what you mean.

> I'd say we want to focus if we have a number. 

Totally agree!

> And we could even reuse the information in ThreadUI.handleActivity instead of having the same "if".

Do you suggest to pass this parameter to "compose" panel args too? So that args for "compose" will change from ```{ activity: ..... }``` to ```{ activity: ....., focusComposer: ....}```.

Comment (julienw) [inline apps/sms/js/activity_handler.js]: yep !
I'm fine if you think this should be in another patch though (because this is some refactoring, not some bug fixing)

Comment (azasypkin) [inline apps/sms/js/activity_handler.js]: Yeah, makes sense, I see that at least ActivityPicker uses ```ActivityHandler.toView({number: ....})```, so that place should be updated too if we want to replace ```if(number)``` with ```if(focusComposer)``` inside ```ThreadUI.handleActivity```. So I'm more confident to do this in the master-only follow-up

**Comment by try-server-hook** [comment]
[azasypkin Aleh Zasypkin (azasypkin)](https://github.com/azasypkin) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=31915fccf805)

**Comment by try-server-hook** [comment]
[azasypkin Aleh Zasypkin (azasypkin)](https://github.com/azasypkin) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=94878e4df937)

**Comment by try-server-hook** [comment]
[azasypkin Aleh Zasypkin (azasypkin)](https://github.com/azasypkin) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=2901787f309c)

**Comment by try-server-hook** [comment]
[azasypkin Aleh Zasypkin (azasypkin)](https://github.com/azasypkin) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=f0e4a17bf1c1)

**Comment by try-server-hook** [comment]
[azasypkin Aleh Zasypkin (azasypkin)](https://github.com/azasypkin) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=3148de3ccfb2)

**Comment by try-server-hook** [comment]
[azasypkin Aleh Zasypkin (azasypkin)](https://github.com/azasypkin) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=2b95acfcd48f)

**Comment by try-server-hook** [comment]
[azasypkin Aleh Zasypkin (azasypkin)](https://github.com/azasypkin) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=42024f5aac0b)
